Firestore/Source/Local/FSTPersistence.h

```diff
@@ -113,26 +114,47 @@ struct FSTTransactionRunner;
 @end
 
 struct FSTTransactionRunner {
-
 // Intentionally disable nullability checking for this function. We cannot properly annotate
 // the function because this function can handle both pointer and non-pointer types. It is an error
 // to annotate non-pointer types with a nullability annotation.
 #pragma clang diagnostic push
 #pragma clang diagnostic ignored "-Wnullability-completeness"
   template <typename F>
-  auto operator() (F block) const -> decltype(block()) {
+  auto operator()(F block) ->
+      typename std::enable_if<std::is_void<decltype(block())>::value, void>::type {
+    __strong id<FSTTransactional> strongDb = _db;
+    if (!strongDb && _expect_db) {
+      FSTCFail(@"Transaction runner accessed without underlying db when it expected one");
+    }
+    if (strongDb) {
+      [strongDb startTransaction];
+    }
+    block();
+    if (strongDb) {
+      [strongDb commitTransaction];
+    }
+  }
+
+  template <typename F>
+  auto operator()(F block) const ->
+      typename std::enable_if<!std::is_void<decltype(block())>::value, decltype(block())>::type {
     using ReturnT = decltype(block());
-    if (_db) {
-      [_db startTransaction];
+    __strong id<FSTTransactional> strongDb = _db;
+    if (!strongDb && _expect_db) {
+      FSTCFail(@"Transaction runner accessed without underlying db when it expected one");
+    }
+    if (strongDb) {
+      [strongDb startTransaction];
     }
     ReturnT result = block();
-    if (_db) {
-      [_db commitTransaction];
+    if (strongDb) {
+      [strongDb commitTransaction];
     }
     return result;
   }
 #pragma clang diagnostic pop
   __weak id<FSTTransactional> _db;
+  bool _expect_db = false;
```

Comment:
This seems like overkill--this can only happen after shutdown and we're not this defensive elsewhere in the system.